src/database-controller/src/core/k8s.js

```diff
@@ -0,0 +1,250 @@
+
+async function getFramework(name, namespace = 'default') {
+  const res = await customObjectsClient.getNamespacedCustomObject(
+    'frameworkcontroller.microsoft.com',
+    'v1',
+    namespace,
+    'frameworks',
+    name,
+  );
+  return res.response;
+}
+
+async function listFramework(name, namespace = 'default') {
+  const res = await customObjectsClient.listNamespacedCustomObject(
+    'frameworkcontroller.microsoft.com',
+    'v1',
+    namespace,
+    'frameworks',
+  );
+  return res.response;
+}
+
+async function createFramework(frameworkDescription, namespace = 'default') {
+  const res = await customObjectsClient.createNamespacedCustomObject(
+    'frameworkcontroller.microsoft.com',
+    'v1',
+    namespace,
+    'frameworks',
+    frameworkDescription,
+  );
+  return res.response;
+}
+
+async function patchFramework(name, data, namespace = 'default') {
+  const res = await customObjectsClient.patchNamespacedCustomObject(
+    'frameworkcontroller.microsoft.com',
+    'v1',
+    namespace,
+    'frameworks',
+    name,
+    data,
+    { headers: { 'Content-Type': 'application/merge-patch+json' } },
+  );
+  return res.response;
+}
+
+async function deleteFramework(name, namespace = 'default') {
+  const res = await customObjectsClient.deleteNamespacedCustomObject(
+    'frameworkcontroller.microsoft.com',
+    'v1',
+    namespace,
+    'frameworks',
+    name,
+  );
+  return res.response;
+}
+
+function getFrameworkInformer(
+  timeoutSeconds = 365 * 86400,
+  namespace = 'default',
+) {
+  /*
+  Usage:
+
+    const informer = getFrameworkInformer()
+    informer.on('add', (obj) => { console.log(`Added: ${obj.metadata.name}`); });
+    informer.on('update', (obj) => { console.log(`Updated: ${obj.metadata.name}`); });
+    informer.on('delete', (obj) => { console.log(`Deleted: ${obj.metadata.name}`); });
+    informer.on('error', (err) => { console.error(err);});
+    informer.start();
+
+  If the informer disconnects normally from API server, it will re-connect automatically.
+  But during the reconnection, listFn will be called again, which is inefficient.
+  According to https://github.com/kubernetes-client/javascript/blob/932c2fbc34db954c6ed397b3cd9ead08b2ff1d10/src/cache.ts#L82-L85,
+  this behavior will be fixed in the future.
+  TO DO: If @kubernetes/client-node fixes this issue, we should upgrade our code to use the new code.
+
+  If the informer encounters any error, it will stop watching, and won't re-connect.
+  One can restart it by informer.on('error', (err) => { informer.start(); }).
+
+  */
+  const listFn = () => {
+    logger.info('Frameworks are listed.');
+    return customObjectsClient.listNamespacedCustomObject(
+      'frameworkcontroller.microsoft.com',
+      'v1',
+      namespace,
+      'frameworks',
+    );
+  };
+  const informer = k8s.makeInformer(
+    kc,
+    `/apis/frameworkcontroller.microsoft.com/v1/frameworks?timeoutSeconds=${timeoutSeconds}`,
+    listFn,
```

Comment:
Why here does not use default namespace

Reply:
see [https://github.com/kubernetes-client/javascript/blob/b218eb9e7f6080239b7061d42a088749b1d4a749/examples/typescript/informer/informer.ts#L11](https://github.com/kubernetes-client/javascript/blob/b218eb9e7f6080239b7061d42a088749b1d4a749/examples/typescript/informer/informer.ts#L11)

---
In reply to: [466166339](https://github.com/microsoft/pai/pull/4761#discussion_r466166339) [](ancestors = 466166339)

Reply:
Previous code is wrong, changed to  `/apis/frameworkcontroller.microsoft.com/v1/namespaces/${namespace}/frameworks?timeoutSeconds=${timeoutSeconds}`

Reply:
So why your previous tests are passed?

Reply:
The url `/apis/frameworkcontroller.microsoft.com/v1/frameworks` also works. It may stand for all frameworks in all namespaces.